File: server/nomadflow/routers/repos.py
"""Repository management endpoints."""

from fastapi import APIRouter, Depends, HTTPException

from ..config import Settings, get_settings
from ..auth import verify_auth
from ..services.git_service import GitService
from ..models.requests import CloneRepoRequest
from ..models.responses import ListReposResponse, CloneRepoResponse

router = APIRouter(prefix="/api", tags=["repos"], dependencies=[Depends(verify_auth)])


def get_git_service(settings: Settings = Depends(get_settings)) -> GitService:
    """Dependency to get GitService instance."""
    return GitService(settings)


@router.post("/list-repos", response_model=ListReposResponse)
async def list_repos(
    git_service: GitService = Depends(get_git_service),
) -> ListReposResponse:
    """
    List all Git repositories in the configured repos directory.

    Returns a list of repositories with their names, paths, and current branches.
    """
    try:
        repos = await git_service.list_repos()
        return ListReposResponse(repos=repos)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/clone-repo", response_model=CloneRepoResponse)
async def clone_repo(
    request: CloneRepoRequest,
    git_service: GitService = Depends(get_git_service),
) -> CloneRepoResponse:
    """Clone a git repository into the repos directory."""
    try:
        name, path, branch = await git_service.clone_repo(
            url=request.url,
            token=request.token,
            name=request.name,
        )
        return CloneRepoResponse(name=name, path=path, branch=branch)
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: nomadflow-rs/crates/nomadflow-server/src/routes/repos.rs
use std::sync::Arc;

use axum::{
    extract::State,
    http::StatusCode,
    routing::post,
    Json, Router,
};
use serde_json::{json, Value};

use nomadflow_core::error::NomadError;
use nomadflow_core::models::{CloneRepoRequest, CloneRepoResponse, ListReposResponse};

use crate::state::AppState;

async fn list_repos(
    State(state): State<Arc<AppState>>,
) -> Result<Json<ListReposResponse>, (StatusCode, Json<Value>)> {
    match state.git.list_repos().await {
        Ok(repos) => Ok(Json(ListReposResponse { repos })),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({ "detail": e.to_string() })),
        )),
    }
}

async fn clone_repo(
    State(state): State<Arc<AppState>>,
    Json(request): Json<CloneRepoRequest>,
) -> Result<Json<CloneRepoResponse>, (StatusCode, Json<Value>)> {
    match state
        .git
        .clone_repo(&request.url, request.token.as_deref(), request.name.as_deref())
        .await
    {
        Ok((name, path, branch)) => Ok(Json(CloneRepoResponse { name, path, branch })),
        Err(NomadError::AlreadyExists(msg)) => Err((
            StatusCode::CONFLICT,
            Json(json!({ "detail": msg })),
        )),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({ "detail": e.to_string() })),
        )),
    }
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/api/list-repos", post(list_repos))
        .route("/api/clone-repo", post(clone_repo))
}
